File: src/main/java/net/gtaun/shoebill/object/impl/ActorImpl.java
package net.gtaun.shoebill.object.impl;

import net.gtaun.shoebill.SampEventDispatcher;
import net.gtaun.shoebill.SampNativeFunction;
import net.gtaun.shoebill.SampObjectStoreImpl;
import net.gtaun.shoebill.data.AngledLocation;
import net.gtaun.shoebill.data.Location;
import net.gtaun.shoebill.data.Vector3D;
import net.gtaun.shoebill.event.destroyable.DestroyEvent;
import net.gtaun.shoebill.exception.CreationFailedException;
import net.gtaun.shoebill.object.Actor;
import net.gtaun.shoebill.object.Player;
import net.gtaun.util.event.EventManager;
import net.gtaun.util.event.EventManagerNode;

/**
 * Created by marvin on 01.05.15 in project shoebill-runtime.
 * Copyright (c) 2015 Marvin Haschker. All rights reserved.
 */
public class ActorImpl implements Actor {
    private int id = INVALID_ACTOR, modelid;
    private EventManagerNode eventManagerNode;

    public ActorImpl(EventManager eventManager, SampObjectStoreImpl store, int modelid, Vector3D pos, float angle) {
        this(eventManager, store, modelid, pos, angle, true, -1);
    }

    public ActorImpl(EventManager eventManager, SampObjectStoreImpl store, int modelid, Vector3D pos, float angle, boolean doExec, int id) {
        this.modelid = modelid;
        this.eventManagerNode = eventManager.createChildNode();

        if (doExec)
            SampEventDispatcher.getInstance().executeWithoutEvent(() -> setup(store, SampNativeFunction.createActor(modelid, pos.x, pos.y, pos.z, angle)));
        else
            setup(store, id);
    }

    private void setup(SampObjectStoreImpl store, int id) {
        if(id == INVALID_ACTOR) throw new CreationFailedException();
        this.id = id;

        store.setActor(id, this);
    }

    @Override
    public int getId() {
        return this.id;
    }

    @Override
    public int getModel() {
        return modelid;
    }

    @Override
    public AngledLocation getLocation() {
        AngledLocation pos = new AngledLocation();
        SampNativeFunction.getActorPos(id, pos);

        pos.angle = SampNativeFunction.getActorFacingAngle(id);
        pos.worldId = SampNativeFunction.getActorVirtualWorld(id);

        return pos;
    }

    @Override
    public int getVirtualWorld() {
        return SampNativeFunction.getActorVirtualWorld(id);
    }

    @Override
    public float getAngle() {
        return SampNativeFunction.getActorFacingAngle(id);
    }

    @Override
    public float getHealth() {
        return SampNativeFunction.getActorHealth(id);
    }

    @Override
    public void setLocation(AngledLocation loc) {
        SampNativeFunction.setActorPos(id, loc.x, loc.y, loc.z);
        SampNativeFunction.setActorVirtualWorld(id, loc.worldId);
        SampNativeFunction.setActorFacingAngle(id, loc.angle);
    }

    @Override
    public void setLocation(Location loc) {
        SampNativeFunction.setActorPos(id, loc.x, loc.y, loc.z);
        SampNativeFunction.setActorVirtualWorld(id, loc.worldId);
    }

    @Override
    public void setLocation(Vector3D pos) {
        SampNativeFunction.setActorPos(id, pos.x, pos.y, pos.z);
    }

    @Override
    public void setAngle(float angle) {
        SampNativeFunction.setActorFacingAngle(id, angle);
    }

    @Override
    public void setVirtualWorld(int world) {
        SampNativeFunction.setActorVirtualWorld(id, world);
    }

    @Override
    public void setHealth(float health) {
        SampNativeFunction.setActorHealth(id, health);
    }

    @Override
    public void setInvulnerable(boolean invulnerable) {
        SampNativeFunction.setActorInvulnerable(id, invulnerable);
    }

    @Override
    public boolean isInvulnerable() {
        return SampNativeFunction.isActorInvulnerable(id) > 0;
    }

    @Override
    public void applyAnimation(String animLib, String animName, float animSpeed, boolean loop, boolean lockX, boolean lockY, boolean freeze, int time) {
        SampNativeFunction.applyActorAnimation(id, animLib, animName, animSpeed, loop ? 1 : 0, lockX ? 1 : 0, lockY ? 1 : 0, freeze ? 1 : 0, time);
    }

    @Override
    public void clearAnimation() {
        SampNativeFunction.clearActorAnimations(id);
    }

    @Override
    public boolean isActorStreamedIn(Player player) {
        return SampNativeFunction.isActorStreamedIn(id, player.getId()) > 0;
    }

    @Override
    public void destroy() {
        if (isDestroyed()) return;

        SampNativeFunction.destroyActor(id);
        destroyWithoutExec();
    }

    @Override
    public boolean isDestroyed() {
        return id == INVALID_ACTOR;
    }

    public void destroyWithoutExec() {
        if (isDestroyed()) return;
        DestroyEvent destroyEvent = new DestroyEvent(this);
        eventManagerNode.dispatchEvent(destroyEvent, this);
        eventManagerNode.destroy();

        this.id = INVALID_ACTOR;
    }
}

File: src/main/java/net/gtaun/shoebill/entities/impl/ActorImpl.kt
package net.gtaun.shoebill.entities.impl

import net.gtaun.shoebill.SampNativeFunction
import net.gtaun.shoebill.SampObjectStoreImpl
import net.gtaun.shoebill.data.AngledLocation
import net.gtaun.shoebill.data.Vector3D
import net.gtaun.shoebill.entities.Actor
import net.gtaun.shoebill.entities.Player
import net.gtaun.shoebill.event.destroyable.DestroyEvent
import net.gtaun.shoebill.exception.CreationFailedException
import net.gtaun.util.event.EventManager
import net.gtaun.util.event.EventManagerNode

/**
 * @author Marvin Haschker
 */
class ActorImpl constructor(eventManager: EventManager, store: SampObjectStoreImpl,
                                          override val model: Int, pos: Vector3D, angle: Float) : Actor() {

    override var id = SampNativeFunction.createActor(model, pos.x, pos.y, pos.z, angle)
        private set

    private val eventManagerNode: EventManagerNode = eventManager.createChildNode()

    init {
        if(id == INVALID_ACTOR) throw CreationFailedException()
        store.setActor(id, this)
    }

    override var location: AngledLocation
        get() {
            val pos = AngledLocation()
            SampNativeFunction.getActorPos(id, pos)

            pos.angle = SampNativeFunction.getActorFacingAngle(id)
            pos.worldId = SampNativeFunction.getActorVirtualWorld(id)

            return pos
        }
        set(loc) {
            SampNativeFunction.setActorPos(id, loc.x, loc.y, loc.z)
            SampNativeFunction.setActorVirtualWorld(id, loc.worldId)
            SampNativeFunction.setActorFacingAngle(id, loc.angle)
        }

    override var virtualWorld: Int
        get() = SampNativeFunction.getActorVirtualWorld(id)
        set(world) {
            SampNativeFunction.setActorVirtualWorld(id, world)
        }

    override var health: Float
        get() = SampNativeFunction.getActorHealth(id)
        set(health) {
            SampNativeFunction.setActorHealth(id, health)
        }

    override var isInvulnerable: Boolean
        get() = SampNativeFunction.isActorInvulnerable(id) > 0
        set(invulnerable) {
            SampNativeFunction.setActorInvulnerable(id, invulnerable)
        }

    override fun applyAnimation(animLib: String, animName: String, time: Int, animSpeed: Float, loop: Boolean,
                                lockX: Boolean, lockY: Boolean, freeze: Boolean) {
        SampNativeFunction.applyActorAnimation(id, animLib, animName, animSpeed, if (loop) 1 else 0,
                if (lockX) 1 else 0, if (lockY) 1 else 0, if (freeze) 1 else 0, time)
    }

    override fun clearAnimation() {
        SampNativeFunction.clearActorAnimations(id)
    }

    override fun isActorStreamedIn(player: Player): Boolean = SampNativeFunction.isActorStreamedIn(id, player.id) > 0

    override fun destroy() {
        if (isDestroyed) return

        SampNativeFunction.destroyActor(id)

        val destroyEvent = DestroyEvent(this)
        eventManagerNode.dispatchEvent(destroyEvent, this)
        eventManagerNode.destroy()

        this.id = Actor.INVALID_ACTOR
    }

    override val isDestroyed: Boolean
        get() = id == Actor.INVALID_ACTOR

}
